### ISSUE TITLE
Re-run failure test to verify JGroups thread findings

### ISSUE THREAD

**Issue by ahus1** [keycloak/keycloak-benchmark#520]
### Description

When running the failure tests #393 and #438 where we killed pods repeatedly during load test runs, we found that reducing the number of Keycloak threads avoided a 'org.jgroups.util.ThreadPool: thread pool is full' message.

This is outlined in https://www.keycloak.org/keycloak-benchmark/kubernetes-guide/latest/load-behavior#threads. 

In a nutshell: If JGroups threads is (by default) configured with 200 threads, a 6-Pod Keycloak cluster should not have more than 33 Keycloak worker threads in a single Keycloak Pod. Keycloak's Helm chart already does that calculation: 

https://github.com/keycloak/keycloak-benchmark/blob/880e4ea05a2262f0dc70861f6c608f17028325ad/provision/minikube/keycloak/templates/keycloak.yaml#L152-L157

The test setup (similar to #393): 

* Single Cluster
* 100_000 clients
* 100_000 users
* 250 logins per second
* 100_000 active sessions (either created beforehand with a long expiry time, or created continuously during the test ... it might be a nice enhancement to have this as a new parameter that N first logins don't log out, and then they log out)
* CPU and memory limits and requests configured to 
* 6 Keycloak pods, 15 DB connections for each pod
* kc-chaos.sh running continuouly
* 5 ec2 instances running the load against the cluster
* OpenShift loadbalancer configured for round-robin
* OTEL enabled
* Checks (including Liveness checks) enabled - the short periods where they fail shouldn't make the pods restart

Hypothesis: 
* When running with the threads limitation, the system will stay stable
  * all nodes will successfully restart
  * there is a small number of rejected requests (503) when threads are blocking during a Keycloak Pod restart
* When running without the threads limitation and each Keycloak Pod running 200 worker threads, 
  * the initial state transfer would sometimes to fail, and Pod restarts take longer
  * the number of rejected requests (503) is larger compared to the first scenario

### Motivation

If this still proves to be valid with the current version of Infinispan, this is something which should be applied Keycloak in the main repository by default. 

If it isn't, it should be removed from the Helm charts and the documentation.

**Comment by martin-kanis** [comment]
I ran first test with the applied change around max number of threads.
Setup https://docs.google.com/spreadsheets/d/15oRZMA72yICu73Hkt1fH4QF-kmKNnvJTBcqojCd6Rq8/edit?pli=1#gid=115045928: 
- 6 pods
- 1 realm, 100k users per realm, 100k clients per realm, 100k active user sessions
- 250 requests per second

Scenario: Extended load test for 4000s, chaos script executed after ~17 minutes

**No org.jgroups.util.ThreadPool: thread pool is full' found in the logs** 

Gatling results:
![Screenshot from 2023-09-11 16-11-35](https://github.com/keycloak/keycloak-benchmark/assets/3359545/6e7035cd-e054-4c8e-a769-ff6fff8348c7)

![Screenshot from 2023-09-11 15-53-08](https://github.com/keycloak/keycloak-benchmark/assets/3359545/2cac0cd6-d2a1-4c50-8abd-ac6b0a994e29)

![Screenshot from 2023-09-11 15-56-42](https://github.com/keycloak/keycloak-benchmark/assets/3359545/6bbf6c4d-b96f-468e-9ce5-28c77cc62ad8)

![Screenshot from 2023-09-11 15-57-17](https://github.com/keycloak/keycloak-benchmark/assets/3359545/72fb5b68-ced2-4643-83f3-1d2fb80cfb62)

![Screenshot from 2023-09-11 16-00-04](https://github.com/keycloak/keycloak-benchmark/assets/3359545/292b01af-2d06-4159-85bc-abb5aa183a59)

![Screenshot from 2023-09-11 16-00-20](https://github.com/keycloak/keycloak-benchmark/assets/3359545/09ff9589-4839-461c-a620-0667560aae05)

**Comment by kami619** [comment]
thats great news, thanks @martin-kanis. One quick question, Were the probe checks (both liveness and readiness) running during the test run ?

**Comment by martin-kanis** [comment]
@ahus1 I tried to run it without `QUARKUS_THREAD_POOL_MAX_THREADS` and with `QUARKUS_THREAD_POOL_QUEUE_SIZE` still in place. I cannot reproduce `org.jgroups.util.ThreadPool: thread pool is full` message.
There are only `Executor thread pool exhausted, starting to reject requests` and `Executor thread pool no longer exhausted, request processing continues after 1 discarded request(s)` messages. Do you thing is worth to remove load shedding as it might mask the `thread pool is full` issue?

PS: I needed to reduce load to 100 users-per-sec as there is CPU throttling with the requested numbers. Even with 100 users-per-sec the cluster is struggling. I might scale up for further tests.

**Comment by martin-kanis** [comment]
Results from first three runs: https://drive.google.com/file/d/1P3p6nrYaUAep3arZpSyyoRKg6w_Vwjt7/view?usp=drive_link

**Comment by ahus1** [comment]
I rerun the test and this is what I found. 
Using the Keycloak nightly build which uses RESTEasy reactive - as it is still blocking on all requests, the results were still similar to the previous tests. The load shedding code isn't present in that version at the moment, so it will log an error message for each dropped request.

* After some debugging, I found that In the original test the message only appeared to the JGroups default at that time that the message was printed when the thread pool was exhausted for the first time. The reason for that value was the same as we were unable to configure the JGroup thread pool size which was 100 at the time and is now back to 200 (see https://issues.redhat.com/browse/ISPN-14780). After setting `-Djgroups.thread_dumps_threshold=1` the message started to appear. 
* The sizing of the environment was not as described in the recipe: for 250 users per second, the recommendation is to have request/limit of 3/9 vCPUs. To allow for a better scheduling in the cluster, I set it to 6/9. 
* Tested the standard configuration with 250 users per second in the steads state and no thread limit -> results were ok 
* Tested with JGroup pool size 10 -> the error message appeared
* Tested with JGroup pool size 100 without the limitation of threads -> the error message appeared when killing pods
* Tested with JGroup pool size 200, increased the users per second to 600 and reduced the number of users per realm  to avoid DB queries which overloaded the system -> error message appeared when killing pods
* Tested as above but with a Keycloak thread size reduced to 200 divided by 6 pod instances -> no error message when killing pods appeared, and the dashboard showing the JGroups thread pool didn't exceed 150 JGroup threads 

The system was still recovering in those scenarios, most likely due to the enabled load shedding and the disabled health checks.

Even though the system recovered, I think we can recommend
* the reduction of Keycloak worker threads to be aligned with the JGroup thread size, as running into retries due to missing threads would still slow down the system
* setting `-Djgroups.thread_dumps_threshold=1` to get notified in the logs when the thread pool is exausted.
* monitoring the metric `vendor_jgroups_tcp_get_thread_pool_size`

Pushed a change making `-Djgroups.thread_dumps_threshold=1` the new default in our Helm charts for Keycloak. 

@pruivo - could you share some rationale behind the default setting of the 10000 in `thread_pool.thread_dumps_threshold="${jgroups.thread_dumps_threshold:10000}`? Should we change this only in Keycloak, or is that something other Infinispan customer should change as well to get an earlier notification if they run into such a situation.

**Comment by belaban** [comment]
@pruivo Such high thread pool numbers can only be achieved when using OOB messages; do we still use OOB? I thought we switched to regular messages.
Also, I assume that we don't use `no-bundler` anymore (`transfer-queue` instead)?

**Comment by ahus1** [comment]
@belaban - Keycloak uses [Infinispan's default Kubernetes configuration](https://github.com/infinispan/infinispan/blob/14.0.17.Final/core/src/main/resources/default-configs/default-jgroups-kubernetes.xml) which sets `bundler_type="transfer-queue"`. I did a short debugging an the setting ends up in the TCP/TP class where it should end up.

To add more context: The exceeded thread pool happend in a scenario where there are 6 Keycloak pods with 200 worker threads each. Each Keycloak pod running an embedded Infinispan with 200 JGroup threads. 

My interpretation of things: When I then disrupt the cluster with starting/stopping Keycloaks under high load, it seems that the 6x200 = 1200 Keycloak worker threads might overwhelm the JGroup thread pool an embedded Infinispan instance that only has 200 JGroup threads.

**Comment by belaban** [comment]
Hi @ahus1 
ack on `transfer-queue`; this is good.

My comment on OOB versus regular was that N regular messages from the same pod use *1* thread, whereas N OOB messages from the same pod use N threads to deliver them up the stack to Infinispan->Keycloak.

When using 6 pods you'd end up with 6+delta (~= 8) threads *max*.

I recall discussions with @pruivo as to why they were using OOB threads which cause a lot of resource usage. Not sure what the outcome was...

**Comment by ahus1** [comment]
@belaban - would there be any multiplier / delta depending on the number of local or distributed caches?

**Comment by belaban** [comment]
I don't think so, as all caches use the same transport, unless I'm mkistaken (@pruivo ?)

**Comment by pruivo** [comment]
Yes, we had a discussion about OOB and Regular messages. We did a PoC and tested it and it is working fine. The only problem is that in Infinispan test suite, the tests testing race condition usually block messages from being delivered which block all the subsequent message. I've fixed a couple of tens of tests but I still have another handful of tests to fix.

The transport is shared among the caches.
